libraries/chain/block_log.cpp

```diff
@@ -286,28 +296,76 @@ namespace eosio { namespace chain {
       }
    }
 
+   std::vector<char>  detail::block_log_impl::pack(const pruned_block& b, pruned_transaction::cf_compression_type segment_compression) {
+
+      // In version 4 of the irreversible blocks log format, these log entries consists of the following in order:
+      //    1. An uint32_t offset from the start of this log entry to the start of the next log entry.
+      //    2. An uint32_t indicating the compression status for the serialization of the pruned_block following this.
+      //    3. The serialization of a pruned_block representation of the block for the entry including padding.
+
+      std::size_t padded_size = b.maximum_pruned_pack_size(segment_compression);
+      std::vector<char> buffer(padded_size + 2* sizeof(uint32_t));
+      fc::datastream<char*> stream(buffer.data(), buffer.size());
+
+      uint32_t offset      = buffer.size();
+      uint32_t compression = static_cast<uint32_t>(segment_compression);
+      stream.write((char*)&offset, sizeof(offset));
+      stream.write((char*)&compression, sizeof(compression));
+      b.pack(stream, segment_compression);
+      return buffer;
+   }
+
+   uint64_t detail::block_log_impl::write_log_entry(const pruned_block& b, pruned_transaction::cf_compression_type segment_compression) {
+      uint64_t pos = block_file.tellp();
+      std::vector<char> buffer;
+     
+      if (version >= 4)  {
+         buffer = detail::block_log_impl::pack(b, segment_compression);
+      } else {
+         auto block_ptr = b.to_signed_block();
```

Comment:
I think it is unfortunate that we need to do a costly conversion to a `signed_block` (and the associated heap allocations) to write out the old blocks to disk. This performance hit isn't as critical as the reads (which can be done multiple times per block rather than just once), but it would still be nice if we did better here. Of course we could do a custom serialization as if it was a signed_block, but I suppose it isn't important to work on right now (and possibly ever because we will encourage people to switch to version 4 block logs). Still, I think we should at least add a `TODO` message here to remind us to go back and improve the performance of this.